packages/@aws-cdk/aws-codepipeline-actions/lib/s3/source-action.ts

```diff
@@ -81,13 +81,34 @@ export class S3SourceAction extends Action {
       outputs: [props.output],
     });
 
+    if (props.bucketKey.length === 0) {
+      throw new Error('Property bucketKey cannot be an empty string');
+    }
+
     this.props = props;
   }
 
   protected bound(_scope: Construct, stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
       codepipeline.ActionConfig {
     if (this.props.trigger === S3Trigger.EVENTS) {
-      this.props.bucket.onCloudTrailPutObject(stage.pipeline.node.uniqueId + 'SourceEventRule', {
+      const baseId = stage.pipeline.node.uniqueId + 'SourceEventRule';
```

Comment:
I think the logic might be easier to follow if you just let `id` be mutable.

```ts
let id = ....;
if (props.bucket.node.tryFindChild(id) !== undefined) {
  id = ...;
}
```

Reply:
`id ` is already mutable: it's defined with `let id = ...;`.